WooCommerce/Classes/POS/Presentation/Barcode Scanning/GameControllerBarcodeParser.swift

```diff
@@ -0,0 +1,197 @@
+import Foundation
+import GameController
+
+/// Parses GameController keyboard input into barcode scans.
+/// This class handles the core logic for interpreting GameController GCKeyCode input as barcode data,
+/// providing language-independent barcode scanning by bypassing iOS keyboard layout processing.
+final class GameControllerBarcodeParser {
+    /// Configuration for the barcode scanner
+    let configuration: HIDBarcodeParserConfiguration
+    /// Callback that is triggered when a barcode scan completes (success or failure)
+    let onScan: (Result<String, Error>) -> Void
+
+    private let timeProvider: TimeProvider
+
+    private var buffer = ""
+    private var lastKeyPressTime: Date?
+
+    init(configuration: HIDBarcodeParserConfiguration,
+         onScan: @escaping (Result<String, Error>) -> Void,
+         timeProvider: TimeProvider = DefaultTimeProvider()) {
+        self.configuration = configuration
+        self.onScan = onScan
+        self.timeProvider = timeProvider
+    }
+
+    /// Process a GameController key press event
+    /// - Parameters:
+    ///   - keyCode: The GameController key code that was pressed
+    ///   - isShiftPressed: Whether shift key is currently pressed
+    func processKeyPress(_ keyCode: GCKeyCode, isShiftPressed: Bool = false) {
+        guard shouldRecogniseAsScanKeystroke(keyCode, isShiftPressed: isShiftPressed) else {
+            return
+        }
+
+        guard let character = characterForKeyCode(keyCode, isShiftPressed: isShiftPressed) else {
+            return
+        }
+
+        if configuration.terminatingStrings.contains(character) {
+            processScan()
+        } else {
+            guard !excludedKeyCodes.contains(keyCode) else { return }
+            checkForTimeoutBetweenKeystrokes()
+            buffer.append(character)
+        }
+    }
+
+    private func shouldRecogniseAsScanKeystroke(_ keyCode: GCKeyCode, isShiftPressed: Bool) -> Bool {
+        guard let character = characterForKeyCode(keyCode, isShiftPressed: isShiftPressed) else {
+            // This prevents a double-trigger-pull on a Star scanner from adding an error row –
+            // Star use this as a shortcut to switch to the software keyboard. They send keycode 174 0xAE, which is
+            // undefined and reserved in UIKeyboardHIDUsage. The scanner doesn't send a character with the code.
+            // There seems to be no reason to handle empty input when considering scans.
+            return false
+        }
+
+        guard character.isNotEmpty else {
+            // This prevents a double-trigger-pull on a Star scanner from adding an error row –
+            // Star use this as a shortcut to switch to the software keyboard. They send keycode 174 0xAE, which is
+            // undefined and reserved in UIKeyboardHIDUsage. The scanner doesn't send a character with the code.
+            // There seems to be no reason to handle empty input when considering scans.
+            return false
```

Comment:
Do we need this comment block to be repeated? Perhaps it could all be in a single guard.